Added more documentation

RichardLitt opened the pull request:
So far: 
- [x] add

RichardLitt committed:
Added Group add

RichardLitt commented on apiary.apib:
```diff
@@ -4,9 +4,41 @@ FORMAT: 1A
+    + chunker (boolean, optional) - chunking algorithm to use
+    + pin (boolean, optional) - Pin this object when adding.  Default true
+
++ Response 200 (text/plain)
```
I was unable to get "http://127.0.0.1:5001/api/v0/add?arg=/Users/richard/test" to work on Postman, so I honestly am not sure if this is the right response of not. It's what I get in the CLI, though.

dignifiedquire replied:
Path to a file will not work from http api, you have to send the file content via multipart if you want to add something via http

mappum replied:
If you need to test adds in Postman, you can use the `form-data` body Content-Type, set a non-empty value for the key, and select 'File' instead of 'Text'.

In curl, you can do:

```
curl -F "file=@/some/file/path" http://localhost:5001/api/v0/add
```

mappum replied:
BTW, could be cool to show `curl` commands for all these methods, does apiary have any good places where we could put that?

RichardLitt replied:
There's no option to select File instead of Text if I am using a GET command. If I use a PUT or a POST, there is, but I just get a `403 - Forbidden` back.  Ideas?

RichardLitt replied:
Looked around: I can't see any way to actually have the curl commands in these.

RichardLitt replied:
We could include the curl command directly in the description, like this. I like it! What do you think? cc @Dignifiedquire 

```
# Group add [GET /add{?arg}{&recursive,quiet,silent,progress,trickle,only%2Dhash,wrap%2Dwith%2Ddirectory,hidden,chunker,pin}]
Add an object to ipfs.

Adds contents of <path> to ipfs. Use -r to add directories.
Note that directories are added recursively, to form the ipfs
MerkleDAG. A smarter partial add with a staging area (like git)
remains to be implemented.

#### curl

    curl -F "file=@test/" -r http://localhost:5001/api/v0/add

+ Parameters
    + arg (string, required) - The path to a file to be added to IPFS
...
```

RichardLitt replied:
Looks like this: 
<img width="687" alt="screenshot 2015-12-25 20 50 09" src="https://cloud.githubusercontent.com/assets/910753/12004833/1fda22da-ab49-11e5-8b56-6dc72ad4a49a.png">

daviddias replied:
> from @RichardLitt: There's no option to select File instead of Text if I am using a GET command. If I use a PUT or a POST, there is, but I just get a 403 - Forbidden back. Ideas?

That would be because no one (i.e Apiary) is expecting for a file to be sent in a GET request and all the other methods are blocked by CORS, if the headers are not set.

RichardLitt commented on apiary.apib:
```diff
@@ -4,9 +4,41 @@ FORMAT: 1A
+    + wrap-with-directory (boolean, optional) - Wrap files with a directory object
+    + hidden (boolean, optional) - Include files that are hidden
+    + chunker (boolean, optional) - chunking algorithm to use
+    + pin (boolean, optional) - Pin this object when adding.  Default true
```
Not sure how to add shortcuts we use here, like -r for --recursive. Should I put them in the desc?

jbenet replied:
good question. desc works? not sure. others' thoughts?

dignifiedquire replied:
Is there actually a need to have these shortcuts working/documented in the http api? Or should they simply be a bonus specific to the cli?

RichardLitt replied:
I don't think we need them.

daviddias replied:
let's make the alias be only available through the CLI :)

RichardLitt replied:
:+1:

RichardLitt committed:
Added block put

RichardLitt committed:
Added config

RichardLitt committed:
Added ipfs daemon

RichardLitt committed:
Added dht

RichardLitt committed:
Added ipfs file

RichardLitt committed:
Added ipfs get

RichardLitt committed:
Filled out ipfs id

RichardLitt committed:
Added in args to calls

RichardLitt committed:
Added ipfs log

RichardLitt committed:
Added ipfs ls

RichardLitt committed:
Added content to mount

RichardLitt committed:
Added ipfs name as much as possible

RichardLitt committed:
Added ipfs object

This one, in particular, is not great, but this should be a good start. There's a few things in the `ipfs object` docs that are inconsistent; for instance, I'm not sure that <name>  and <ref> aren't options. The "--encoding" flag is mentioned several times, but never specified as an option, unless it is in `put`, where it has the name `inputenc`, which I think does something different. The formatting is also a bit inconsistent - tick marks given here but not there, and so on. I think I'll need to comb over this in more detail later. For now, here's a stub. Better to have something than nothing.

RichardLitt committed:
Added `ipfs pin` group

It seems strange to me that pin/rm has the same 200 response as pin/add. But that seems to be the case.

RichardLitt committed:
Added options and args to ipfs ping

RichardLitt committed:
Added `ipfs refs`

I got a malformed syntax error when trying refs on Postman, but it seemed to work fine in the command line. I am not sure what exactly is happening with the format option, and I suspect that coud be improved a bit. Otherwise, this seems fine to me. Threw in headers because I might as well, they were right there.

RichardLitt committed:
Added params to resolve

RichardLitt committed:
Added params to stats bw

RichardLitt committed:
Added content to ipfs swarm

RichardLitt committed:
Added `ipfs tar`

I didn't get a response in my terminal for tarcat, so I assume there is none. I was unable to get tar/add to work, as well, in Postman, but in the terminal it returned the hash of the file, so that is probably what it returns there, too.

RichardLitt committed:
Added params to version

RichardLitt commented:
Ok, I've finished going through all of the commands for the first time. Now is the harder bit; which is, namely, figuring out why some don't work in the API at all, where the inconsistencies are in the documentation, and how to clean up and make this, as a whole, better. For now, though, I think this is a massive step forward. Can someone give a brief overview of my work here, point out any glaring issues?

RichardLitt commented:
This, of course, relates and possibly closes #7.

jbenet commented:
cc @whyrusleeping @diasdavid @lgierth @mappum for review

RichardLitt committed:
Linted according to API doc

RichardLitt committed:
Remove test file

RichardLitt committed:
Changed nullable to string, nullable

RichardLitt commented:
I have now linted this, and it works in apiary. Whether or not it is 100% accurate is up for debate; but, this means we have something to work with, as far as three columns and apiary goes.

Dignifiedquire commented:
Okay, I will start reviewing this and then we can merge this for now and have any bikeshedding discussions later, as the most important thing for now is to get the IS state documented

dignifiedquire commented on apiary.apib:
```diff
@@ -79,8 +116,29 @@ Store data as a raw ipfs block.
+It reads from stdin, and <key> is a base58 encoded multihash.
+
++ Parameters
+    + arg (string, required) - The path to the data to be stored as an IPFS block
```
I don't think this works, `file paths` do not work on the http api as far as I know, or at least there is also an alternative way of uploading the data directly

RichardLitt replied:
It works with curl. Added example response and curl request.

dignifiedquire commented on apiary.apib:
```diff
@@ -244,41 +305,301 @@ Lists all available commands (and subcommands) that are available.
+
+Set the value of the 'datastore.path' key:
+
+    `ipfs config datastore.path ~/.ipfs/datastore`
```
These examples should not be here, as they document the cli

dignifiedquire commented on apiary.apib:
```diff
@@ -244,41 +305,301 @@ Lists all available commands (and subcommands) that are available.
+## config [GET /config{?arg1,arg2}{&bool,json}]
+
+ipfs config controls configuration variables. It works
+much like 'git config'. The configuration values are stored in a config
```
please use ticks instead of '

dignifiedquire commented on apiary.apib:
```diff
@@ -244,41 +305,301 @@ Lists all available commands (and subcommands) that are available.
+    + arg1: "Addresses.API" (string, required) - The key of the config entry
+    + arg2 (string, required) - The value to set the config entry to
+    + bool (boolean, optional) - Set a boolean
+    + json (boolean, optional) - Parse stringified JSON
```
It's not documented how you would actually send json

dignifiedquire commented on apiary.apib:
```diff
@@ -244,41 +305,301 @@ Lists all available commands (and subcommands) that are available.
+
++ Parameters
+
+    + arg (string, required) - The file to use as the new config
```
Will this work with a file path, or do you need to attach the actual content in the request?

RichardLitt replied:
It works with a file path. Added a `curl` example, and example headers. This should be a 204, not a 200. See #14.

dignifiedquire commented on apiary.apib:
```diff
@@ -244,41 +305,301 @@ Lists all available commands (and subcommands) that are available.
+documented in (and can be modified through) `ipfs config Addresses`.
+For example, to change the `Gateway` port:
+
+    ipfs config Addresses.Gateway /ip4/127.0.0.1/tcp/8082
```
Again please no cli examples in here

dignifiedquire commented on apiary.apib:
```diff
@@ -449,91 +868,370 @@ ipfs.io. TXT "dnslink=/ipfs/QmRzTuh2Lpuz7Gr39stNr6mTFdqAghsZec1JoUnfySUzcy ..."
+Publish an object to IPNS
+
++ Parameters
+    + arg (string, required) - IPFS path of the obejct to be published
```
typo `object`

dignifiedquire commented on apiary.apib:
```diff
@@ -449,91 +868,370 @@ ipfs.io. TXT "dnslink=/ipfs/QmRzTuh2Lpuz7Gr39stNr6mTFdqAghsZec1JoUnfySUzcy ..."
+`ipfs object put` is a plumbing command for storing DAG nodes.
+It reads from stdin, and the output is a base58 encoded multihash.
+
+Data should be in the format specified by the --inputenc flag.
```
Please no cli speak like `--flag`

dignifiedquire commented on apiary.apib:
```diff
@@ -449,91 +868,370 @@ ipfs.io. TXT "dnslink=/ipfs/QmRzTuh2Lpuz7Gr39stNr6mTFdqAghsZec1JoUnfySUzcy ..."
+## patch [GET /object/patch{?arg1,arg2}{&name,ref}]
+Create a new merkledag object based on an existing one
+
+`ipfs object patch <root> <cmd> <args>` is a plumbing command used to
```
again, this is cli notation

RichardLitt replied:
Yeah, I don't know about this one. I wasn't able to get it to work on Postman. Ought we to remove it?

dignifiedquire commented on apiary.apib:
```diff
@@ -449,91 +868,370 @@ ipfs.io. TXT "dnslink=/ipfs/QmRzTuh2Lpuz7Gr39stNr6mTFdqAghsZec1JoUnfySUzcy ..."
+## rm [GET /pin/rm{?arg}{&recursive}]
+Removes the pinned object from local storage.
+
+By default, recursively. Use `-r=false` for direct pins.
```
cli notation

dignifiedquire commented on apiary.apib:
```diff
@@ -563,10 +1261,54 @@ trip latency information.
+
+  <link base58 hash>
+
+Note: list all refs recursively with -r.
```
cli notation

dignifiedquire commented on apiary.apib:
```diff
@@ -563,10 +1261,54 @@ trip latency information.
+Retrieves the object named by <ipfs-path> and displays the link
+hashes it contains, with the following format:
+
+  <link base58 hash>
```
strange indentation

mappum commented on apiary.apib:
```diff
@@ -4,9 +4,42 @@ FORMAT: 1A
+remains to be implemented.
+
++ Parameters
+    + arg (string, required) - The path to a file to be added to IPFS
```
We might want to indicate that there can be multiple `arg` values to add multiple files in one request

ABISprotocol replied:
@mappum Excuse me, I realize this is OT for here, but just a brief inquiry, do you have a status or roadmap for Mercury?

mappum replied:
@ABISprotocol Not sure why you're bringing that up here. My email address is on my profile, feel free to contact me there. :smile:

mappum commented on apiary.apib:
```diff
@@ -79,8 +116,29 @@ Store data as a raw ipfs block.
+Stores input as an IPFS block
+
+ipfs block put is a plumbing command for storing raw ipfs blocks.
+It reads from stdin, and <key> is a base58 encoded multihash.
```
Stdin isn't relevant in a HTTP API

RichardLitt committed:
Single quotes to backticks

RichardLitt committed:
Update add with new curl result

This is not the result I got back from Postman, as I wasn't able to make it work there. Here is the curl response.

RichardLitt committed:
Using a data structure for error messages

RichardLitt committed:
Remove CLI docs

RichardLitt committed:
Added example curl requests

RichardLitt committed:
Filled out block/put example

RichardLitt committed:
Added curl and headers and 500 to config/replace

RichardLitt committed:
Spelling

RichardLitt committed:
Remove cli speak

RichardLitt committed:
Update refs with better responses

RichardLitt commented:
@Dignifiedquire Want to take another pass?

RichardLitt committed:
Edited dredd.yml to match @Dignifiedquire

RichardLitt commented:
Removed 'needs review' pending current changes ongoing after the API talk.

RichardLitt committed:
Filled out the ipfs add further

RichardLitt commented:
Please do not close or delete this branch. I am using it to stub out future PRs. This PR will not and should not be merged.
